src/lib/notifier.c

```diff
@@ -14,82 +14,175 @@
 
-static struct notify_data _notify_data __aligned(PLATFORM_DCACHE_ALIGN);
-
-void notifier_register(struct notifier *notifier)
+#define trace_notifier(__e, ...) \
+	trace_event(TRACE_CLASS_NOTIFIER, __e, ##__VA_ARGS__)
+#define tracev_notifier(__e, ...) \
+	tracev_event(TRACE_CLASS_NOTIFIER, __e, ##__VA_ARGS__)
+#define trace_notifier_error(__e, ...) \
+	trace_error(TRACE_CLASS_NOTIFIER, __e, ##__VA_ARGS__)
+
+struct notify_data {
+	void *caller;
+	enum notify_id type;
+	uint32_t data_size;
+	void *data;
+} __aligned(PLATFORM_DCACHE_ALIGN);
+
+static struct notify_data _notify_data[PLATFORM_CORE_COUNT]
+	__aligned(PLATFORM_DCACHE_ALIGN);
+
+struct callback_handle {
+	void *receiver;
+	void *caller;
+	void (*cb)(void *arg, enum notify_id, void *data);
+	struct list_item list;
+};
+
+int notifier_register(void *receiver, void *caller, enum notify_id type,
+		       void (*cb)(void *arg, enum notify_id type, void *data))
 {
 	struct notify *notify = *arch_notify_get();
+	struct callback_handle *handle;
+
+	if (type >= NOTIFIER_ID_MAX) {
+		trace_notifier_error("notifier_register() error: invalid ID to "
+				     "register to.");
+		return -EINVAL;
+	}
+
+	handle = rzalloc(RZONE_SYS_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*handle));
+
+	if (!handle) {
+		trace_notifier_error("notifier_register() error: callback "
+				     "handle allocation failed.");
+		return -ENOMEM;
+	}
+
+	list_init(&handle->list);
+	handle->receiver = receiver;
+	handle->caller = caller;
+	handle->cb = cb;
 
 	spin_lock(notify->lock);
-	list_item_prepend(&notifier->list, &notify->list);
+	list_item_prepend(&handle->list, &notify->list[type]);
 	spin_unlock(notify->lock);
+
+	return 0;
 }
 
-void notifier_unregister(struct notifier *notifier)
+void notifier_unregister(void *receiver, void *caller, enum notify_id type)
 {
 	struct notify *notify = *arch_notify_get();
+	struct list_item *wlist;
+	struct list_item *tlist;
+	struct callback_handle *handle;
+
+	if (type >= NOTIFIER_ID_MAX) {
+		trace_notifier_error("notifier_unregister() error: invalid ID "
+				     "to unregister from.");
+		return;
+	}
 
+	/*
+	 * Unregister all matching callbacks
+	 * If receiver is NULL, unregister all callbacks with matching callers
+	 * If caller is NULL, unregister all callbacks with matching receivers
+	 *
+	 * Event producer might force unregister all receivers by passing
+	 * receiver NULL
+	 * Event consumer might unregister from all callers by passing caller
+	 * NULL
+	 */
 	spin_lock(notify->lock);
-	list_item_del(&notifier->list);
+	list_for_item_safe(wlist, tlist, &notify->list[type]) {
+		handle = container_of(wlist, struct callback_handle, list);
+		if ((!receiver || handle->receiver == receiver) &&
+		    (!caller || handle->caller == caller)) {
+			list_item_del(&handle->list);
+			rfree(handle);
+		}
+	}
 	spin_unlock(notify->lock);
 }
 
-void notifier_notify(void)
+void notifier_unregister_all(void *receiver, void *caller)
+{
+	int i;
+
+	for (i = 0; i < NOTIFIER_ID_ALL; i++)
```

Comment:
I think that's mistake. MAX should be used in cases like this (iterators) and ALL in case when you want to broadcast, they have the same value now but they don't have to.

Reply:
@jajanusz Shouldn't ALL then get some sort of reserved value like -1?

Reply:
@paulstelian97 This comes from @jajanusz comment https://github.com/thesofproject/sof/pull/1950#discussion_r336417600 , but after @tlauda suggested to implement unsubscribe_all function, I believe that the 'ALL' enum value is obsolete. I'm going to remove it is a moment.